src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2024 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// the corresponding header
+#include "dbgroup/atomic/mwcas/lock_free/mwcas_descriptor.hpp"
+
+// C++ standard libraries
+#include <atomic>
+#include <cstddef>
+
+// external libraries
+#include "dbgroup/lock/common.hpp"
+#include "dbgroup/memory/utility.hpp"
+
+// local sources
+#include "dbgroup/atomic/mwcas/utility.hpp"
+
+//                       Bit allocation of a word.
+// |     63     |       62-50       |      49-47     |        46-0       |
+// | MwCAS Flag | Reference Counter | Begin Position |Descriptor Address |
+
+namespace dbgroup::atomic::mwcas::lock_free
+{
+namespace
+{
+/*##############################################################################
+ * Local constants
+ *############################################################################*/
+
+/// @brief An offset for right-shifting to extract the "begin position".
+constexpr uint64_t kPosShift = 47;
+
+/// @brief An offset for right-shifting to extract the "reference counter".
+constexpr uint64_t kCntShift = 50;
+
+/// @brief A constant for incrementing the "reference counter".
+constexpr uint64_t kCntUnit = 1UL << kCntShift;
+
+/// @brief A bitmask with only the "begin position" portion set to 1.
+constexpr uint64_t kPosMask = (((1UL << (1 + kCntShift - kPosShift)) - 1) << kPosShift);
+
+/// @brief A bitmask with only the "reference counter" portion set to 1.
+constexpr uint64_t kCntMask = (((1UL << (64 - kCntShift)) - 1) << kCntShift);
+
+/// @brief A bitmask with only the "MwCAS FLAG" and "descriptor address" portions set to 1.
+constexpr uint64_t kDescMask = kMwCASFlag | MwCASDescriptor::kAddrMask;
+
+/*##############################################################################
+ * Local global variable
+ *############################################################################*/
+thread_local std::unique_ptr<MwCASDescriptor> tls = nullptr;
```

Comment:
Doxygenコメントを追加しました．また，これはあんまり一般的ではないですが，グローバル変数を区別するために接頭辞として `_` を入れておいてください．

```suggestion

/// @brief A thread local descriptor for reuse.
thread_local std::unique_ptr<MwCASDescriptor> _tls = nullptr;  // NOLINT
```